src/local/room_name.rs

```diff
@@ -0,0 +1,346 @@
+use std::{
+    error,
+    fmt::{self, Write},
+    ops,
+    str::FromStr,
+};
+
+use arrayvec::ArrayString;
+
+/// A structure representing a room name.
+#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug)]
+pub struct LocalRoomName {
+    /// Inner x coordinate representation.
+    ///
+    /// 0 represents E0, positive numbers represent E(x)
+    ///
+    /// -1 represents W0, negative numbers represent W((-x) - 1)
+    pub x_coord: i32,
+    /// Inner y coordinate representation.
+    ///
+    /// 0 represents S0, positive numbers represent S(y)
+    ///
+    /// -1 represents N0, negative numbers represent N((-y) - 1)
+    pub y_coord: i32,
+}
+
+impl fmt::Display for LocalRoomName {
+    /// Formats this room name into the format the game expects.
+    ///
+    /// Resulting string will be `(E|W)[0-9]+(N|S)[0-9]+`, and will result
+    /// in the same LocalRoomName if passed into [`LocalRoomName::new`].
+    ///
+    /// [`LocalRoomName::new`]: struct.LocalRoomName.html#method.new
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if self.x_coord >= 0 {
+            write!(f, "E{}", self.x_coord)?;
+        } else {
+            write!(f, "W{}", (-self.x_coord) - 1)?;
+        }
+
+        if self.y_coord >= 0 {
+            write!(f, "S{}", self.y_coord)?;
+        } else {
+            write!(f, "N{}", (-self.y_coord) - 1)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl LocalRoomName {
+    /// Creates a new room name from the given input.
+    ///
+    /// This will parse the input, and return an error if it is in an invalid
+    /// format.
+    #[inline]
+    pub fn new<T>(x: &T) -> Result<Self, LocalRoomNameParseError>
+    where
+        T: AsRef<str> + ?Sized,
+    {
+        x.as_ref().parse()
+    }
+
+    /// Creates a new room name from the given position parameters.
+    #[inline]
+    pub fn from_coords(east: bool, south: bool, x_pos: i32, y_pos: i32) -> Self {
+        LocalRoomName {
+            x_coord: if east { x_pos } else { -x_pos - 1 },
+            y_coord: if south { y_pos } else { -y_pos - 1 },
+        }
+    }
+
+    /// Converts this LocalRoomName into an efficient, stack-based string.
+    ///
+    /// This is equivalent to [`ToString::to_string`], but involves no
+    /// allocation.
+    pub fn to_array_string(&self) -> ArrayString<[u8; 8]> {
+        let mut res = ArrayString::new();
+        write!(res, "{}", self).expect("expected ArrayString write to be unfallible");
+        res
+    }
+}
+
+impl ops::Add<(i32, i32)> for LocalRoomName {
+    type Output = Self;
+
+    /// Adds an (x, y) coordinate pair to this room name.
```

Comment:
This again assumes that users are aware of the implementation.

Reply:
Documentation added.